app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
     end
+
+    @remaining_hierarchy = hierarchy
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
+  def lock_key
+    first_argument = arguments.first
+    first_argument.is_a?(String) ? first_argument : Random.uuid
   end
 
   private
 
-  def called_from_provider_hierarchy?
-    return unless (tenant_id = object.tenant_id)
-    caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
+  def hierarchy_entries_for(...)
+    self.class.send(:hierarchy_entries_for, ...)
   end
 
-  def callback_options
-    { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
-  end
-
-  class ReflectionDestroyer
-
-    def initialize(main_object, reflection, caller_worker_hierarchy)
-      @main_object = main_object
-      @reflection = reflection
-      @caller_worker_hierarchy = caller_worker_hierarchy
-    end
-
-    def destroy_later
-      reflection.many? ? destroy_has_many_association : destroy_has_one_association
+  # handles a single hierarchy entry
+  # @return [Array<String>] hierarchy for a newly discovered object from association to delete or empty array otherwise
+  def handle_one_hierarchy_entry!(hierarchy)
+    entry = hierarchy.pop
+    case entry
+    when /Plain-([:\w]+)-(\d+)/
+      ar_object = $1.constantize.find($2.to_i)
+      match = hierarchy.last&.match(ASSOCIATION_RE)
+      if match
+        # callbacks logic differs between object destroyed by association, so set it here
+        # e.g. for Plans, to prevent acts_as_list to update position of each plan on deletion
+        # we want to have here a reflection with a foreign key that is part of the list scope array.
+        # see acts_as_list/active_record/acts/scope_method_definer.rb
+        # Note that if you hand-craft garbage hierarchy, it may still count as the expected foreighn key,
+        # e.g. you delete an ApplicationPlan but association is Account:account_plans or
+        # e.g. the association is for an unrelated parent object id
+        # FYI if there is no such association (e.g. a method name), nil will be returned.
+        # Another possible confusion with hand-crafter hierarchies may occur when before
+        # a Plain- entry, there is set an unrelated Hierarchy- entry, then incorrect association will be set.
+        ar_object.destroyed_by_association = match[:klass].constantize.reflect_on_association(match[:association])
+      end
+      ar_object.background_deletion_method_call
+    when ASSOCIATION_RE
+      hierarchy.concat handle_association($1.constantize.find($2.to_i), $3, entry)
+    else
+      raise ArgumentError, "Invalid entry specification: #{entry}"
     end
+    hierarchy
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
+    []
```

Comment:
So, we ignore the rest of the hierarchy if the parent object is already gone?... 

or, maybe I'm just confused :grimacing:

Reply:
Yes. Correct. Now you're asking, I see how one may try to obtain any one-layer objects associated with the deleted one. But can't look any deeper. I think it would be a needless complication at this point to try handle one layer of associated objects of a missing parent.


If we have some broken remains of hierarchies, I think it makes more sense to detect orphan objects from removed tenants and delete them within a separate worker, than making a bigger mess here that will not ensure full cleanup anyway.

Hope this makes sense. Let me know if it needs further elaboration.